nibabel/cmdline/diff.py

```diff
@@ -0,0 +1,121 @@
+
+import sys
+from optparse import OptionParser, Option
+
+import numpy as np
+
+import nibabel as nib
+import nibabel.cmdline.utils
+from nibabel.cmdline.utils import _err, verbose, table2string, ap, safe_get
+import fileinput
+
+__author__ = 'Yaroslav Halchenko & Christopher Cheng'
+__copyright__ = 'Copyright (c) 2017 NiBabel contributors'
+__license__ = 'MIT'
+
+
+def get_opt_parser():
+    # use module docstring for help output
+    p = OptionParser(
+        usage="%s [OPTIONS] [FILE ...]\n\n" % sys.argv[0] + __doc__,
+        version="%prog " + nib.__version__)
+
+    p.add_options([
+        Option("-v", "--verbose", action="count",
+               dest="verbose", default=0,
+               help="Make more noise.  Could be specified multiple times"),
+
+        Option("-H", "--header-fields",
+               dest="header_fields", default='all',
+               help="Header fields (comma separated) to be printed as well (if present)"),
+    ])
+
+    return p
+
+
+def diff_files(key, inputs):
+    diffs = []
+
+    for i in range(len(inputs)):
+        if i != len(inputs)-1:
+            temp_input_1 = inputs[i]
+            temp_input_2 = inputs[i+1]
+            if np.any(temp_input_1[key] != temp_input_2[key]):
+                if i != 0 and temp_input_2 != diffs:
+                    diffs.append(temp_input_1[key])
+                    diffs.append(temp_input_2[key])
+                else:
+                    diffs.append(temp_input_1[key])
+            elif type(temp_input_1[key]) != type(temp_input_2[key]):
+                if i != 0 and temp_input_2 != diffs:
+                    diffs.append(temp_input_1[key])
+                    diffs.append(temp_input_2[key])
+                else:
+                    diffs.append(temp_input_1[key])
+            else:
+                pass
+
+    # TODO: figure out a way to not have these erroneous outputs occur in the above loop
+    for a in range(len(diffs)-1):
+        for b in range(len(diffs)-1):
+            try:
+                if np.all(np.isnan(diffs[a])) and np.all(np.isnan(diffs[a+1])):
+                    del diffs[a]
+            except TypeError:
+                pass
+            if a and b != len(diffs)-1:
+                if np.any(diffs[a] == diffs[b]):
+                    del diffs[a]
+
+    if len(diffs) > 1:
+        return {key: diffs}
+
+
+def process_files(files, opts):
+
+    header_list = []
+
+    for f in files:
+        header_list.append(nib.load(f).header)
+
+    if opts.header_fields:
+        # signals "all fields"
+        if opts.header_fields == 'all':
+            # TODO: header fields might vary across file types, thus prior sensing would be needed
+            header_fields = header_list[0].keys()
+        else:
+            header_fields = opts.header_fields.split(',')
+
+        for f in header_fields:
+            if diff_files(f, header_list) is not None:
+                    print(diff_files(f, header_list))
```

Comment:
you might better off again just returning a dict with those header fields where the diff was found.  This would allow for testing it independently of how you decide to output them at the end

Reply:
and then you could rename the function from `process-files` to a more descriptive `get_headers_diff`

Reply:
and why to invoke the same function twice to do the same? assign to some variable.

Also you could try to make these 3 lines into a dict comprehension (see above comments about what the function might return instead ;))